### PR TITLE
Deprecate Ordinary database and old *MergeTree syntax

### LINKED ISSUE #38312: Setting to forbid creation of Ordinary databases
Ordinary databases are deprecated in favor of Atomic databases. The setting will help to enforce usage of the recommended database engine Atomic.

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
`Ordinary` database engine and old storage definition syntax for `*MergeTree` tables are deprecated. By default it's not possible to create new ones. If `system` database has `Ordinary` engine it will be automatically converted to `Atomic` on server startup. There are settings to keep old behavior (`allow_deprecated_database_ordinary` and `allow_deprecated_syntax_for_merge_tree`), but these settings may be removed in future releases.  

Resolves #38312

### CODE REVIEW

**Comment by filimonov** [comment]
Can we add a check to system warnings if filesystem / kernel support `renameat2`?

Without that (examples: zfs, Centos 7 +xfs) the Atomic does not work properly.

**Comment by tavplubix** [comment]
> Without that (examples: zfs, Centos 7 +xfs) the Atomic does not work properly.

It may not support `EXCHANGE TABLES` and other extensions, but it should work properly. Please create an issue if it does not.

**Comment by tavplubix** [comment]
Hm, it's funny, i just typed "centos 7" in google search and the first suggestion was "centos 7 end of life".

**Comment by tavplubix** [comment]
But anyway it makes sense to add a check for kernel version/support of some syscalls to the list of sanity checks on startup.

**Comment by filimonov** [comment]
> centos 7 end of life

still 2 years :D 

> It may not support EXCHANGE TABLES and other extensions

Yes, and (more painful): REPLACE DICTIONARY

File: src/Interpreters/loadMetadata.cpp [hunk 1]
```diff
@@ -201,6 +204,110 @@ static void loadSystemDatabaseImpl(ContextMutablePtr context, const String & dat
     }
 }
 
+static void convertOrdinaryDatabaseToAtomic(ContextMutablePtr context, const DatabasePtr & database)
```
Comment (filimonov) [inline src/Interpreters/loadMetadata.cpp:207]: Maybe we can make some command for that? `Alter database engine` or `system` something?

Comment (filimonov) [reply src/Interpreters/loadMetadata.cpp:207]: Example https://kb.altinity.com/engines/altinity-kb-atomic-database-engine/how-to-convert-ordinary-to-atomic/

Comment (tavplubix) [reply src/Interpreters/loadMetadata.cpp:207]: > Maybe we can make some command for that? Alter database engine or system something?

Yes, I will add it later. The problem is that it's much easier (and safer) to convert database on server startup. Maybe we can create a file in `flags/` directory on "alter database engine" query and actually convert it on next restart.

Comment (tavplubix) [reply src/Interpreters/loadMetadata.cpp:207]: > Example https://kb.altinity.com/engines/altinity-kb-atomic-database-engine/how-to-convert-ordinary-to-atomic/

I have better example =)
https://github.com/ClickHouse/ClickHouse/blob/880780c7a8e4f20afeb06987a48afc7e8dbbf4ae/tests/queries/0_stateless/01155_rename_move_materialized_view.sql

File: docker/test/stress/run.sh [hunk 2]
```diff
@@ -284,6 +284,11 @@ then
 
     rm -rf /var/lib/clickhouse/*
 
+    # Make BC check more funny by forcing Ordinary engine for system database
```
Comment (alesapin) [inline docker/test/stress/run.sh:287]: image

File: src/Databases/DatabaseFactory.cpp [hunk 3]
```diff
@@ -138,8 +121,13 @@ DatabasePtr DatabaseFactory::getImpl(const ASTCreateQuery & create, const String
         throw Exception(ErrorCodes::BAD_ARGUMENTS, "Database engine `{}` cannot have table overrides", engine_name);
 
     if (engine_name == "Ordinary")
+    {
+        if (!create.attach && !context->getSettingsRef().allow_deprecated_database_ordinary)
+            throw Exception(ErrorCodes::UNKNOWN_DATABASE_ENGINE, "Ordinary database engine is deprecated");
```
Comment (alesapin) [inline src/Databases/DatabaseFactory.cpp:126]: Maybe give a hint to user how to return it back?

File: src/Interpreters/loadMetadata.cpp [hunk 4]
```diff
@@ -208,6 +210,110 @@ static void loadSystemDatabaseImpl(ContextMutablePtr context, const String & dat
...
+{
+    /// It's kind of C++ script that creates temporary table with Atomic engine,
```
Comment (alesapin) [inline src/Interpreters/loadMetadata.cpp:215]:
```suggestion
    /// It's kind of C++ script that creates temporary database with Atomic engine,
```

File: src/Storages/MergeTree/registerStorageMergeTree.cpp [hunk 5]
```diff
@@ -372,17 +384,11 @@ static StoragePtr create(const StorageFactory::Arguments & args)
             throw Exception("Expected two string literal arguments: zookeeper_path and replica_name", ErrorCodes::BAD_ARGUMENTS);
 
         /// Allow implicit {uuid} macros only for zookeeper_path in ON CLUSTER queries
-        bool is_on_cluster = args.getLocalContext()->getClientInfo().query_kind == ClientInfo::QueryKind::SECONDARY_QUERY;
```
Comment (alesapin) [inline src/Storages/MergeTree/registerStorageMergeTree.cpp:375]: why we don't need this check no more?

File: programs/server/Server.cpp [hunk 6]
```diff
@@ -1488,6 +1488,7 @@ int Server::main(const std::vector<std::string> & /*args*/)
         /// We load temporary database first, because projections need it.
         database_catalog.initializeAndLoadTemporaryDatabase();
         loadMetadataSystem(global_context);
+        maybeConvertOrdinaryDatabaseToAtomic(global_context, DatabaseCatalog::instance().getSystemDatabase());
```
Comment (alesapin) [inline programs/server/Server.cpp:1491]: maybe make this conversion optional?

File: src/Storages/MergeTree/registerStorageMergeTree.cpp [hunk 7]
```diff
@@ -304,12 +304,24 @@ static StoragePtr create(const StorageFactory::Arguments & args)
                             arg_idx, e.message(), getMergeTreeVerboseHelp(is_extended_storage_def));
         }
     }
+    else if (!args.attach && !args.local_context.lock()->getSettingsRef().allow_deprecated_syntax_for_merge_tree)
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "This syntax for *MergeTree engine is deprecated. "
```
Comment (alesapin) [inline src/Storages/MergeTree/registerStorageMergeTree.cpp:309]: add a hint to allow it?

**Review by alesapin — APPROVED** [review]
Actually there are not too much for review and changes are alright. But fixing outdated tests is big a work.

File: programs/server/Server.cpp (hunk 6 above)
Comment (tavplubix) [reply programs/server/Server.cpp:1491]: It is optional: 
https://github.com/ClickHouse/ClickHouse/blob/94dd80e90069bca17066abadb3bf60d3ff496e59/src/Interpreters/loadMetadata.cpp#L282-L283
(I don't want to introduce another setting for it)

**Comment by tavplubix** [comment]
Stateless tests (aarch64, actions) - some infrastructure issue, cc: @Felixoid  
```
docker: error during connect: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/containers/create": EOF.
See 'docker run --help'.
```

Integration tests flaky check (asan, actions) - actually all tests are either PASSED or SKIPPED, but runner was killed by timeout - OK
```
[gw0] [ 95%] PASSED test_server_initialization/test.py::test_partially_dropped_tables 
test_server_initialization/test.py::test_live_view_dependency 
[gw0] [ 95%] PASSED test_server_initialization/test.py::test_live_view_dependency 
test_system_merges/test.py::test_merge_simple[] 
[gw0] [ 96%] PASSED test_system_merges/test.py::test_merge_simple[] 
test_system_merges/test.py::test_merge_simple[replicated] Killed
```

Performance tests - `Unknown setting allow_deprecated_syntax_for_merge_tree` - not sure how to avoid it, will try to add new setting first

Stateless tests (address, actions) [2/2] - need to update tests (but it's strange that it's the only failed check)

Stress test (undefined, actions) - #38220

**Comment by tavplubix** [comment]
> Stateless tests (address, actions) [2/2] - need to update tests (but it's strange that it's the only failed check)

```
Traceback (most recent call last):
  File "functional_test_check.py", line 315, in <module>
    state, description, test_results, additional_logs = process_results(
  File "functional_test_check.py", line 166, in process_results
    test_results = list(csv.reader(results_file, delimiter="\t"))
_csv.Error: field larger than field limit (131072)
```

https://github.com/ClickHouse/ClickHouse/pull/38030#issuecomment-1168557441

**Comment by azat** [comment]
>_csv.Error: field larger than field limit (131072)

Interesting, I've looked at this report and I don't see csv row > 128K there:

```python
In [14]: max(map(lambda line: max(map(len, line)), csv.reader(open("test_results.tsv", "r"), delimiter="\t")))
Out[14]: 5385
```

**Comment by tavplubix** [comment]
@azat, you looked at wrong report, see https://github.com/ClickHouse/ClickHouse/runs/7085200751?check_suite_focus=true (no report was actually generated, because task has failed because of this error)

**Comment by tavplubix** [comment]
@Mergifyio update

**Comment by tavplubix** [comment]
AST fuzzer (UBSan, actions) - #14700
Integration tests flaky check (asan, actions) - actually it's timeout, no failed tests - OK
Stateless tests (thread, actions) [3/3] - tests with tsan are still broken after upgrading to clang-14
Stateless tests flaky check (address, actions) - `01233_check_table_with_metadata_cache` is flaky, cc: @taiyang-li, @CurtizJ

File: docker/test/stress/run.sh (hunk 2 above)
Comment (zhanglistar) [reply docker/test/stress/run.sh:287]: @alesapin Is that you?

**Comment by zhanglistar** [comment]
@tavplubix I suggest that leave the legacy ordinary database code to allow user old ordinary database in system, but not allowed to create new ordinary databases. Here is the reason, we have a cluster updated from 20.03 to 21.10, there are lots of legacy ordinary databases, if new version do not support ordinary databases, then we can not update version anymore newer.

**Comment by taiyang-li** [comment]
Do we have any advice for old CH users imgrating from ordinary engine to atomic engine gracefully?  Thanks

**Comment by tavplubix** [comment]
> I suggest that leave the legacy ordinary database code to allow user old ordinary database in system, but not allowed to create new ordinary databases. 

@zhanglistar, please read the changelog entry in the PR description. That's exactly what is done in this PR. We do not remove Ordinary engine now and are not going to remove it in the near future.

**Comment by tavplubix** [comment]
> Do we have any advice for old CH users imgrating from ordinary engine to atomic engine gracefully? Thanks

```
CREATE DATABASE db_tmp ENGINE=Atomic;
RENAME TABLE db.table1 TO db_tmp.table1;
RENAME TABLE db.table2 TO db_tmp.table2;
...
DROP DATABASE db;
RENAME DATABASE db_tmp TO db;
```

See also example in test: https://github.com/ClickHouse/ClickHouse/blob/880780c7a8e4f20afeb06987a48afc7e8dbbf4ae/tests/queries/0_stateless/01155_rename_move_materialized_view.sql
And code that converts `system` database: https://github.com/ClickHouse/ClickHouse/blob/57686b842226b3a65c042a341b48105f4919ec9d/src/Interpreters/loadMetadata.cpp#L214-L276

A query that converts database engine will be added: https://github.com/ClickHouse/ClickHouse/pull/38335#discussion_r905905501

**Comment by mfridman** [comment]
Hey @tavplubix, is there an ENV variable one can set the equivalent of `allow_deprecated_syntax_for_merge_tree` ?

**Comment by den-crane** [comment]
@mfridman try use `from_env="..."` 

export MY_ENV_VAR=1

```xml
cat /etc/clickhouse-server/users.d/ENV_NAME.xml
<?xml version="1.0" ?>
<yandex>
    <profiles>
        <default>
            <allow_deprecated_syntax_for_merge_tree  from_env="MY_ENV_VAR"/>
        </default>
    </profiles>
</yandex>
```

**Comment by mfridman** [comment]
> @mfridman try to use `from_env="..."`
> 
> export MY_ENV_VAR=1
> 
> ```
> cat /etc/clickhouse-server/users.d/ENV_NAME.xml
> <?xml version="1.0" ?>
> <yandex>
>     <profiles>
>         <default>
>             <allow_deprecated_syntax_for_merge_tree  from_env="MY_ENV_VAR"/>
>         </default>
>     </profiles>
> </yandex>
> ```

Thanks, I invoke the vanilla container from [clickhouse/clickhouse-server](https://hub.docker.com/r/clickhouse/clickhouse-server/) without passing any options. Was hoping there would be some way to pass ENV variables directly to the container without dealing with config files. But afaics that's not possible.

**Comment by victor-perov** [comment]
Thanks for working on this, @tavplubix!

Would you mind shedding a bit more light on what do you think the timeline for _allow_deprecated_database_ordinary_?
We are interested in keeping **Ordinary** up until we can do the conversion properly (see issue #42250). So I'm curious what is let's say an EOL for Ordinary engine? Can we expect it to be around for v23?

Kind regards,
Viktor

**Comment by tavplubix** [comment]
There's no EOL, we are not going to remove the `Ordinary` engine in the near future. However, we are not going to fix minor bugs and usability issues in `Ordinary` either. Some new features may not work with `Ordinary` too. We will start to show a warning about `Ordinary` after https://github.com/ClickHouse/ClickHouse/issues/42250 is fixed and maybe will make `allow_deprecated_database_ordinary` harder to enable.

**Comment by victor-perov** [comment]
Thanks, @tavplubix!

**Pull request merged**